pkg/agent/agent_loop_test.go

```diff
@@ -0,0 +1,313 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package agent
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubectl-ai/gollm"
+	"github.com/GoogleCloudPlatform/kubectl-ai/pkg/api"
+	"github.com/GoogleCloudPlatform/kubectl-ai/pkg/tools"
+)
+
+type mockChatResponse struct {
+	gollm.ChatResponse
+	candidates []gollm.Candidate
+}
+
+func (r *mockChatResponse) Candidates() []gollm.Candidate {
+	return r.candidates
+}
+
+func (r *mockChatResponse) UsageMetadata() any {
+	return nil
+}
+
+type mockCandidate struct {
+	gollm.Candidate
+	parts []gollm.Part
+}
+
+func (c *mockCandidate) Parts() []gollm.Part {
+	return c.parts
+}
+
+func (c *mockCandidate) String() string {
+	var result string
+	for _, part := range c.parts {
+		if text, ok := part.AsText(); ok {
+			result += text
+		}
+	}
+	return result
+}
+
+type mockTextPart struct {
+	gollm.Part
+	text string
+}
+
+func (p *mockTextPart) AsText() (string, bool) {
+	return p.text, true
+}
+
+func (p *mockTextPart) AsFunctionCalls() ([]gollm.FunctionCall, bool) {
+	return nil, false
+}
+
+type mockFunctionCallPart struct {
+	gollm.Part
+	calls []gollm.FunctionCall
+}
+
+func (p *mockFunctionCallPart) AsText() (string, bool) {
+	return "", false
+}
+
+func (p *mockFunctionCallPart) AsFunctionCalls() ([]gollm.FunctionCall, bool) {
+	return p.calls, true
+}
+
+// newMockChatResponse creates a gollm.ChatResponseIterator from a list of parts.
+func newMockChatResponse(parts ...gollm.Part) gollm.ChatResponseIterator {
+	return func(yield func(gollm.ChatResponse, error) bool) {
+		response := &mockChatResponse{
+			candidates: []gollm.Candidate{
+				&mockCandidate{
+					parts: parts,
+				},
+			},
+		}
+		yield(response, nil)
+	}
+}
+
+func TestAgentRun_LLMTextResponse(t *testing.T) {
+	t.Parallel()
+	mockChat := &mockChat{
+		sendStreamingFunc: func(ctx context.Context, contents ...any) (gollm.ChatResponseIterator, error) {
+			return newMockChatResponse(&mockTextPart{text: "hello from the mock"}), nil
+		},
+	}
+	a := newTestAgent(mockChat)
+	a.RunOnce = true
+
+	err := a.Run(context.Background(), "test query")
+	if err != nil {
+		t.Fatalf("Run() error = %v", err)
+	}
+
+	// Expect a text response from the model
+	timeout := time.After(1 * time.Second)
+	var receivedResponse bool
+	for !receivedResponse {
+		select {
+		case msg := <-a.Output:
+			m := msg.(*api.Message)
+			if m.Type == api.MessageTypeText && m.Source == api.MessageSourceModel {
+				if strings.Contains(m.Payload.(string), "hello from the mock") {
+					receivedResponse = true
+				}
+			}
+		case <-timeout:
+			t.Fatal("timed out waiting for model response")
+		}
+	}
+
+	for range a.Output {
+		if a.AgentState() != api.AgentStateExited {
+			t.Errorf("Expected agent state to be Exited, but got %s", a.AgentState())
+		}
+	}
+}
+
+func TestAgentRun_LLMErrorResponse(t *testing.T) {
```

Comment:
I think TestAgentRun_LLMTextResponse and TestAgentRun_LLMErrorResponse can be combined as table driven form because the driver code is almost same.